### PR TITLE
【Hackathon 6th No.30】为 paddle.nn.functional.max_pool1d/max_pool2d /max_pool3d/paddle.signal.stft 进行功能增强

### PR DESCRIPTION
**Pull request by AndPuQing**
### PR Category
User Experience

### PR Types
New features

### Description

1. 修复了在`return_mask=True`模式下，`ceil_mode`失效的问题，并添加了相应单测。

2. ~~修复了 stft 在 `as_real` 后出现的数据不一致问题~~(develop分支无法复现)

image

另外，stft 是否需要根据输入数据 x 所在的设备来决定输出结果的设备，develop分支中，即使 `x` 在 cpu 上，输出tensor 依然在gpu上(因为window 创建时是根据default_device决定的)

### CODE REVIEW

File: python/paddle/signal.py [hunk 1]
```diff
@@ -362,6 +362,7 @@ def stft(
         ), f'expected a 1D window tensor of size equal to win_length({win_length}), but got window with shape {window.shape}.'
     else:
         window = paddle.ones(shape=(win_length,), dtype=x.dtype)
+        window = paddle.to_tensor(window, place=x.place)
```
Comment (zhwesky2010) [inline python/paddle/signal.py]: 这里windows的place为何是错的，一般情况下无需针对输出设备进行设置的

File: python/paddle/signal.py [hunk 2]
```diff
@@ -362,6 +362,7 @@ def stft(
...
```
Comment (zhwesky2010) [inline python/paddle/signal.py]: 正常情况下：API的输入与输出是在相同设备上的，无需指定，底层执行器会自动设定相应的place

File: test/legacy_test/test_pool1d_api.py [hunk 3]
```diff
@@ -296,6 +296,22 @@ def check_avg_dygraph_padding_same(self, place):
...
+        with base.dygraph.guard(place):
+            input_np = np.random.random([1, 3, 6]).astype("float32")
+            input = paddle.to_tensor(input_np)
+            result, _ = F.max_pool1d(
+                input, kernel_size=5, ceil_mode=True, return_mask=True
+            )
+            result_np = max_pool1D_forward_naive(
+                input_np,
+                ksize=[5],
+                strides=[5],
```
Comment (zhwesky2010) [inline test/legacy_test/test_pool1d_api.py:370]: stride和padding为何设置不一致？

File: test/legacy_test/test_pool1d_api.py [hunk 4]
```diff
@@ -296,6 +296,22 @@ def check_avg_dygraph_padding_same(self, place):
...
+                paddings=[0],
+                ceil_mode=True,
+            )
+            np.testing.assert_allclose(result.numpy(), result_np, rtol=1e-05)
```
Comment (zhwesky2010) [inline test/legacy_test/test_pool1d_api.py:374]: 判断一下：两者的输出shape一致

File: test/legacy_test/test_pool2d_api.py [hunk 5]
```diff
@@ -357,6 +357,22 @@ def check_avg_divisor(self, place):
...
             np.testing.assert_allclose(result.numpy(), result_np, rtol=1e-05)
 
+    def check_max_pool_return_mask_ceil(self, place):
+        with base.dygraph.guard(place):
+            input_np = np.random.random([2, 3, 33, 33]).astype("float32")
+            input = paddle.to_tensor(input_np)
+            result, _ = max_pool2d(
+                input, kernel_size=5, ceil_mode=True, return_mask=True
+            )
+            result_np = pool2D_forward_naive(
```
Comment (zhwesky2010) [inline test/legacy_test/test_pool2d_api.py:375]: 同上

File: test/legacy_test/test_pool3d_api.py [hunk 6]
```diff
@@ -356,6 +356,22 @@ def check_avg_divisor(self, place):
...
             np.testing.assert_allclose(result.numpy(), result_np, rtol=1e-05)
 
+    def check_max_pool_return_mask_ceil(self, place):
+        with base.dygraph.guard(place):
+            input_np = np.random.random([1, 2, 6, 33, 33]).astype("float32")
+            input = paddle.to_tensor(input_np)
+            result, _ = max_pool3d(
+                input, kernel_size=5, ceil_mode=True, return_mask=True
+            )
+            result_np = pool3D_forward_naive(
```
Comment (zhwesky2010) [inline test/legacy_test/test_pool3d_api.py:374]: 同上

**Review by zhwesky2010 — COMMENTED** [review]
1. max_pool1d/2d/3d重点需要检验下在ceil_mode为True/False下的shape是否一致？
2. `paddle.signal.sftf`与`torch.sftf`结果做了对比吗，详细对比两者下两者的结果
3. 设备的问题：正常情况下，API的输入输出的设备是一致的，无需特殊指定配置，这里为何会出现不一致并且要手动指定？

**Comment by zhwesky2010** [comment]
@AndPuQing 另外，该PR合入后，需要提交 API中文文档、API映射文档的修改。

同时需要将PaConvert中 `_test` 屏蔽的case打开，提交一个PR会自动触发测试。

File: python/paddle/signal.py (hunk 1 above)
Comment (AndPuQing) [reply python/paddle/signal.py]: 这里当用户默认设备是GPU，但是输入的`X`指定为CPU上，同时用户没有提供`window`输入，这里会创建一个在默认设备(GPU)上的tensor，最后造成输出时result在GPU，这是符合预期的吗？

**Comment by AndPuQing** [comment]
### Paddle version
```shell
commit: 02eb1dc336cf650ac3b5c5d2fdbf55f893f7fb02
cuda: 12.0
cudnn: 8.9.1
nccl: 21701
xpu: False
xpu_xccl: False
xpu_xhpc: False
cinn: False
```

### Pytorch version
```shell
2.2.2+cu121
39901f229520a5256505ec24782f716ee7ddc843
```
**input dtype = float32**

| input_shape | n_fft | hop_length | normalized | onesided | Max absolute diff. | Max relative diff. |
|:-----------:|:-----:|:----------:|:----------:|:--------:|:------------------:|:------------------:|
|   (2,100)   |   50  |     15     |    False   |   True   |    1.3126562e-06   |    9.709884e-06    |
|   (2,100)   |  100  |     15     |    False   |   True   |    3.8146973e-06   |     9.33037e-06    |
|   (2,500)   |   50  |     15     |    False   |   True   |    1.9073486e-06   |    9.538695e-06    |
|   (2,500)   |  100  |     15     |    False   |   True   |    2.1324806e-06   |    8.827294e-05    |
|   (2,500)   |  200  |     15     |    False   |   True   |    7.6293945e-06   |    8.2466126e-05   |
|   (2,500)   |  200  |     50     |    False   |   True   |    3.8146973e-06   |     5.94125e-05    |
|   (2,500)   |  200  |     15     |    True    |   True   |    4.7683716e-07   |    7.2988776e-05   |
|   (2,2000)   |  400  |     15     |   False    |   True   |    1.5258789e-05   |    0.00011829   |

**input dtype = float64**

| input_shape | n_fft | hop_length | normalized | onesided | Max absolute diff. | Max relative diff. |
|:-----------:|:-----:|:----------:|:----------:|:--------:|:------------------:|:------------------:|
|   (2,100)   |   50  |     15     |    False   |   True   |   3.55271368e-15   |   5.96877696e-15   |
|   (2,100)   |  100  |     15     |    False   |   True   |   7.10542736e-15   |   4.95725428e-14   |
|   (2,500)   |   50  |     15     |    False   |   True   |   3.55271368e-15   |   2.49609664e-14   |
|   (2,500)   |  100  |     15     |    False   |   True   |   7.10542736e-15   |   4.42488659e-14   |
|   (2,500)   |  200  |     15     |    False   |   True   |   1.42108547e-14   |   1.87328036e-13   |
|   (2,500)   |  200  |     50     |    False   |   True   |   9.28696749e-15   |    1.8887612e-13   |
|   (2,500)   |  200  |     15     |    True    |   True   |   1.33226763e-15   |    5.8703342e-14   |
|   (2,2000)  |  400  |     15     |    False   |   True   |   1.77635684e-15   |   2.74193926e-13   |

### Test Code

```python
import paddle.cuda_env
import paddle.version
import torch
import paddle
import numpy as np
import torch.version

paddle.version.show()
print(torch.version.__version__)
print(torch.version.git_version)
n_fft = 400
hop_length = 15
normalized = False
onesided = True
inputs_np = np.random.rand(2, 2000).astype('float32')
window_np = np.hamming(n_fft).astype('float32')

# Paddle
x = paddle.to_tensor(inputs_np)
window = paddle.to_tensor(window_np)
res = paddle.signal.stft(
    x=x,
    n_fft=n_fft,
    hop_length=hop_length,
    window=window,
    normalized=normalized,
    onesided=onesided,
)

res_a = paddle.as_real(res)

# PyTorch
x = torch.tensor(inputs_np)
window = torch.tensor(window_np)
res0 = torch.stft(
    input=x,
    n_fft=n_fft,
    hop_length=hop_length,
    window=window,
    normalized=normalized,
    onesided=onesided,
    return_complex=True,
)

res_b = torch.view_as_real(res0)

# Compare results
np.testing.assert_allclose(
    res.numpy(), res0.numpy(), rtol=1e-20, atol=1e-20, verbose=True
)
```

**Comment by zhwesky2010** [comment]
@AndPuQing 
1. 第1个问题修复没问题
2. 第2个问题验证无误，无需修改了
3. 第3个问题，输出out与输入x是应该要保持place一致的，这个原因是`paddle.to_tensor` 默认会使用全局place，因此导致中间结果window的place与输入x不一致；这个问题需要修一下，另外这个文件中的其他window也有这个问题，统一修改下吧

**Comment by luotao1** [comment]
> 第1个问题修复没问题

也可以单独提PR先把问题1给合入了 @AndPuQing

**Comment by AndPuQing** [comment]
> @AndPuQing 另外，该PR合入后，需要提交 API中文文档、API映射文档的修改。
> 
> 同时需要将PaConvert中 `_test` 屏蔽的case打开，提交一个PR会自动触发测试。

1. 不知道这个文档指的什么文档呢？因为ceil_mode原本就有这个api. 
2. 问题3 我另外单独提PR

**Comment by zhwesky2010** [comment]
> > @AndPuQing 另外，该PR合入后，需要提交 API中文文档、API映射文档的修改。
> > 同时需要将PaConvert中 `_test` 屏蔽的case打开，提交一个PR会自动触发测试。
> 
> 1. 不知道这个文档指的什么文档呢？因为ceil_mode原本就有这个api.
> 2. 问题3 我另外单独提PR

@AndPuQing 
- API中文文档需要修改：https://github.com/PaddlePaddle/docs 你修改了API，对应的API介绍文档，例如参数介绍 这些都需要对应修改到与Paddle repo一致
- API映射文档需要修改：https://github.com/PaddlePaddle/docs/tree/develop/docs/guides/model_convert/convert_from_pytorch/api_difference 找到对应的Pytorch API，API参数发生变化就需要修改对应的映射关系，比如与torch参数之间的对应关系发生变化

**Review by zhwesky2010 — DISMISSED** [review]

**Review by zhwesky2010 — COMMENTED** [review]
在 `paddle/phi/ops/yaml/op_version.yaml` 里需要修改下

action:
    add_attr

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by yuanlehome — APPROVED** [review]

**Review by zyfncg — APPROVED** [review]
